selfservice/strategy/password/registration_test.go

```diff
@@ -587,43 +587,77 @@ func TestRegistration(t *testing.T) {
+					values := func(v url.Values) {
+						v.Set("traits.username", "registration-identifier-8-spa-duplicate-"+suffix)
+						v.Set("password", x.NewUUID().String())
+						v.Set("traits.foobar", "bar")
+					}
 
-			t.Run("type=spa", func(t *testing.T) {
-				values := func(v url.Values) {
-					v.Set("traits.username", "registration-identifier-8-spa-duplicate")
-					v.Set("password", x.NewUUID().String())
-					v.Set("traits.foobar", "bar")
-				}
+					_ = expectSuccessfulLogin(t, false, true, nil, values)
+					body := expectValidationError(t, false, true, applyTransform(values, transform))
+					assert.Contains(t, gjson.Get(body, "ui.messages.0.text").String(), "An account with the same identifier (email, phone, username, ...) exists already.", "%s", body)
+				})
+
+				t.Run("type=browser", func(t *testing.T) {
+					values := func(v url.Values) {
+						v.Set("traits.username", "registration-identifier-8-browser-duplicate-"+suffix)
+						v.Set("password", x.NewUUID().String())
+						v.Set("traits.foobar", "bar")
+					}
 
-				_ = expectSuccessfulLogin(t, false, true, nil, values)
-				body := expectValidationError(t, false, true, values)
-				assert.Contains(t, gjson.Get(body, "ui.messages.0.text").String(), "An account with the same identifier (email, phone, username, ...) exists already.", "%s", body)
+					_ = expectSuccessfulLogin(t, false, false, nil, values)
+					body := expectValidationError(t, false, false, applyTransform(values, transform))
+					assert.Contains(t, gjson.Get(body, "ui.messages.0.text").String(), "An account with the same identifier (email, phone, username, ...) exists already.", "%s", body)
+				})
+			}
+
+			t.Run("case=identical input", func(t *testing.T) {
+				testWithTransform(t, "identical", func(v url.Values) {
+					// base case
+				})
 			})
 
-			t.Run("type=browser", func(t *testing.T) {
-				values := func(v url.Values) {
-					v.Set("traits.username", "registration-identifier-8-browser-duplicate")
-					v.Set("password", x.NewUUID().String())
-					v.Set("traits.foobar", "bar")
-				}
+			t.Run("case=different capitalization", func(t *testing.T) {
+				testWithTransform(t, "caps", func(v url.Values) {
+					v.Set("traits.username", strings.ToUpper(v.Get("traits.username")))
+				})
+			})
 
-				_ = expectSuccessfulLogin(t, false, false, nil, values)
-				body := expectValidationError(t, false, false, values)
-				assert.Contains(t, gjson.Get(body, "ui.messages.0.text").String(), "An account with the same identifier (email, phone, username, ...) exists already.", "%s", body)
+			t.Run("case=leading whitespace", func(t *testing.T) {
+				testWithTransform(t, "leading", func(v url.Values) {
+					v.Set("traits.username", "  "+v.Get("traits.username"))
+				})
+			})
+
+			t.Run("case=trailing whitespace", func(t *testing.T) {
+				testWithTransform(t, "trailing", func(v url.Values) {
+					v.Set("traits.username", v.Get("traits.username")+"  ")
+				})
```

Comment:
So you added the two transforms that add whitespace in the beginning and end, but there is no more test without adding whitespace? Can you also add a noop transform to keep the current test behavior?

Reply:
The no-op case still exists of course, it can be found starting at line 639. I apologize if this setup is too confusing, it was the best I could come up with to not duplicate lots and lots of code...

Reply:
I see, sorry. It is a good approach, the diff is just hard to read.